src/gluonts/model/deepar/_schema.py

```diff
@@ -0,0 +1,149 @@
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        raise NotImplementedError()
+
+    def is_compatible(self, v: Any) -> bool:
+        raise NotImplementedError()
+
+
+@dataclass
+class NumpyArrayField(FieldType[np.ndarray]):
+    dtype: Type = np.float32
+    ndim: Optional[int] = None
+
+    def __eq__(self, other):
+        if not isinstance(other, NumpyArrayField):
+            return False
+
+        return self.dtype == other.dtype and self.ndim == other.ndim
+
+    def __call__(self, value: Any) -> np.ndarray:
+        value = np.asarray(value, dtype=object)
+        value = value.astype(self.dtype)
+        if self.ndim is not None and self.ndim != value.ndim:
+            raise GluonTSDataError(
+                f"expected array with dimension {self.ndim}, but got {value.ndim}."
+            )
+        return value
+
+    def is_compatible(self, value: Any) -> bool:
+        if not isinstance(value, (list, tuple, np.ndarray)):
+            return False
+
+        try:
+            x = np.asarray(value)
+        except Exception:
+            return False
+
+        # int types
+        if self.dtype in [np.int32, np.int64]:
+            if x.dtype.kind != "i":
+                return False
+            return self.ndim is None or x.ndim == self.ndim
+
+        # float types
+        try:
+            x = np.asarray(value, dtype=self.dtype)
+        except Exception:
+            return False
+        return self.ndim is None or x.ndim == self.ndim
+
+
+Freq = Union[str, pd.DateOffset]
+
+
+class PandasPeriodField(FieldType[pd.Period]):
+    def __init__(
+        self,
+        freq: Freq,
+    ) -> None:
+        self.freq = to_offset(freq) if isinstance(freq, str) else freq
+
+    def __eq__(self, other):
+        if not isinstance(other, PandasPeriodField):
+            return False
+        return self.freq == other.freq
+
+    def __repr__(self):
+        return f"PandasPeriodField(freq={self.freq!r})"
+
+    def __call__(self, value: Any) -> pd.Period:
+        return pd.Period(value, freq=self.freq)
+
+    def is_compatible(self, v: Any) -> bool:
+        if not isinstance(v, (str, pd.Period)):
+            return False
+        try:
+            self(v)
+        except Exception:
+            return False
+        return True
+
+
+class Schema:
+    def __init__(self, fields: Dict[str, FieldType]) -> None:
+        self.fields = fields
+
+    def __eq__(self, other):
+        return self.fields == other.fields
+
+    def __repr__(self):
+        return (
+            "Schema(fields={"
+            + ", ".join(f"'{k}':{v}" for k, v in self.fields.items())
+            + "})"
+        )
+
+    def __call__(self, d: Dict[str, Any], inplace: bool):
+        """
+        inplace
+            applies the schema to a data dict if True. The dictionary is updated in place.
+            return a new data dictionary if False.
+        """
+        copied_data: Dict[str, Any] = dict()
```

Comment:
You can do something like:

```suggestion
        if inplace:
            out = d
        else:
            out: Dict[str, Any] = {}
```

And then always write to `out`.